iceoryx2-ffi/ffi/src/node.rs

```diff
@@ -0,0 +1,348 @@
+// Copyright (c) 2024 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache Software License 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
+// which is available at https://opensource.org/licenses/MIT.
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+#![allow(non_camel_case_types)]
+
+use super::{IntoCInt, IOX2_OK};
+
+use iceoryx2::prelude::*;
+use iceoryx2::{node::NodeCreationFailure, service};
+use iceoryx2_bb_elementary::static_assert::*;
+
+use core::ffi::c_int;
+use core::mem::{align_of, size_of, MaybeUninit};
+use std::alloc::{alloc, dealloc, Layout};
+
+// TODO: [#210] Add structs iox2_node_name_storage_internal_t, iox2_node_name_storage_t and iox2_node_name_h
+// TODO: [#210] Add API iox2_node_name_new and iox2_node_name_set
+
+// BEGIN iox2_node_builder_* types
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub enum iox2_node_creation_failure_e {
+    INSUFFICIENT_PERMISSIONS = 1, // start at 1 since IOX2_OK is already 0
+    INTERNAL_ERROR,
+}
+
+impl IntoCInt for NodeCreationFailure {
+    fn into_c_int(self) -> c_int {
+        (match self {
+            NodeCreationFailure::InsufficientPermissions => {
+                iox2_node_creation_failure_e::INSUFFICIENT_PERMISSIONS
+            }
+            NodeCreationFailure::InternalError => iox2_node_creation_failure_e::INTERNAL_ERROR,
+        }) as c_int
+    }
+}
+
+#[repr(C)]
+#[repr(align(8))] // alignment of NodeBuilder
+pub struct iox2_node_builder_storage_internal_t {
+    internal: [u8; 18416], // magic number obtained with size_of::<NodeBuilder>()
+}
+
+#[repr(C)]
+pub struct iox2_node_builder_storage_t {
+    internal: iox2_node_builder_storage_internal_t,
+    deleter: fn(*mut iox2_node_builder_storage_t),
+}
+
+/// The handle to use for the `iox2_node_builder_*` functions
+pub type iox2_node_builder_h = *mut iox2_node_builder_storage_t;
+
+impl iox2_node_builder_storage_t {
+    fn node_builder_maybe_uninit(&mut self) -> &mut MaybeUninit<NodeBuilder> {
+        static_assert_gt_or_equal::<
```

Comment:
This can be easily forgotten, when we add - for whatever reason - another init call. I understand why you prefer this over tests, since the user on an unknown platform may not run them. But when we extend the API, this would be the first thing I would forget and then the user runs into undefined behavior.

So I would prefer some additional tests that fail when I adjust the NodeBuilder and forget to add the C size/alignment.

Reply:
Well, we already have a test which runs through this code path, so adding it to another `init` function would be discouraged anyway. There should always be only one place with the static assert and this code path should be tested, which is also already the case. With the current setup I cannot happen to overlook a change in the `NodeBuilder` or `Node` size and alignment. But if it helps you to sleep more peaceful, I can refactor the check into a separate method and have a test calling only that method.